File: ruby/resistor-color-trio/resistor_color_trio.rb
=begin
Write your code for the 'Resistor Color Trio' exercise in this file. Make the tests in
`resistor_color_trio_test.rb` pass.

To get started with TDD, see the `README.md` file in your
`ruby/resistor-color-trio` directory.
=end

class ResistorColorTrio
  attr_reader :colors

  COLORS = %w[black brown red orange yellow green blue violet grey white]

  def initialize(colors)
    @colors = colors
    
  end

  def label
    raise ArgumentError unless all_colors_valid?

    if ohms.length > 3
      display_value = ohms[0..-4]
      display_unit = "kiloohms"
    else
      display_value = ohms
      display_unit = "ohms"
    end

    "Resistor value: #{display_value} #{display_unit}"
  end

  private

  def all_colors_valid?
    colors.all? { |color| COLORS.include?(color) }
  end

  def first_two_digits
    colors.first(2).map { |color| COLORS.index(color) }
  end

  def number_of_zeroes
    COLORS.index(colors[2])
  end

  def ohms
    first_two_digits.join("") + ("0" * number_of_zeroes)
  end
end

File: typescript/resistor-color-trio/resistor-color-trio.ts
export function decodedResistorValue(bands: [string, string, string]): string {
  let firstDigit = colorToValue(bands[0]);
  let secondDigit = colorToValue(bands[1]);
  let numZeroes = colorToValue(bands[2]);

  return resistorValue(firstDigit, secondDigit, numZeroes);
};

function colorToValue(color: string): number {
  let colorMapping: { [key: string]: number } = {
    'black': 0,
    'brown': 1,
    'red': 2,
    'orange': 3,
    'yellow': 4,
    'green': 5,
    'blue': 6,
    'violet': 7,
    'grey': 8,
    'white': 9
  };

  return colorMapping[color];
};

function resistorValue(firstDigit: number, secondDigit: number, numZeroes: number): string {
  let result: string = firstDigit.toString();

  if (secondDigit == 0) {
    numZeroes += 1;
  } else {
    result += secondDigit.toString();
  }

  let zeroes = "0".repeat(numZeroes % 3);

  return result + zeroes + " " + resistanceUnit(numZeroes);
};

function resistanceUnit(numZeroes: number): string {
  if (numZeroes >= 3) {
    return "kiloohms"
  } else {
    return "ohms"
  }
};
